### PR TITLE
planner, executor: support intersection type IndexMerge with hint in planner

### LINKED ISSUE #39335: Support IndexMerge Intersection (planner part)
## Feature Request
sub task of https://github.com/pingcap/tidb/issues/39333
**Is your feature request related to a problem? Please describe:**

**Describe the feature you'd like:**

**Describe alternatives you've considered:**

**Teachability, Documentation, Adoption, Migration Strategy:**

### PR DESCRIPTION
**Pull request by time-and-fate**
### What problem does this PR solve?

Issue Number: close #39335

Problem Summary:

### Check List

Tests 

- [ ] Unit test
- [ ] Integration test
- [ ] Manual test (add detailed scripts or steps below)
- [ ] No code

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

Documentation

- [ ] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features
- [ ] Changes MySQL compatibility

### Release note

Please refer to [Release Notes Language Style Guide](https://pingcap.github.io/tidb-dev-guide/contribute-to-tidb/release-notes-style-guide.html) to write a quality release note.

```release-note
None
```

### CODE REVIEW

**Review by chrysan — APPROVED** [review]

File: planner/core/stats.go [hunk 1]
```diff
@@ -754,6 +786,103 @@ func (ds *DataSource) buildIndexMergeOrPath(filters []expression.Expression, par
...
+			if _, ok := notCoveredHashCodeSet[hashCode]; !ok {
+				hashCodeSet[hashCode] = struct{}{}
+			}
+		}
+
+		finalFilters = append(finalFilters, notCoveredConds...)
+		partialFilters = append(partialFilters, coveredConds...)
+	}
+
+	// Remove covered filters from finalFilters and deduplicate finalFilters.
```
[the remark this answers is not in the document]
Comment (time-and-fate) [reply planner/core/stats.go:870]: Actually, we already have a `RemoveDupExprs([]Expression)` function. But the use case here is a bit special, we use the hashcode set built from the logic above.
So I think we don't need to add another function for this.

File: planner/core/testdata/index_merge_suite_out.json [hunk 2]
```diff
@@ -9,5 +9,438 @@
...
+          "      ├─IndexRangeScan(Build) 9.00 cop[tikv] table:t1, index:ibc(b, c) range:[20,20], keep order:false",
+          "      └─Selection(Probe) 3.60 cop[tikv]  gt(test.t1.e, test.t2.a)",
+          "        └─TableRowIDScan 4.50 cop[tikv] table:t1 keep order:false"
+        ],
+        "Result": [
+          "20",
+          "20",
+          "20"
+        ],
+        "Warning": null
```
[the remark this answers is not in the document]
Comment (time-and-fate) [reply planner/core/testdata/index_merge_suite_out.json]: Updated.
I deleted all warnings in the JSON and asserted no warnings in the test logic.

**Review by guo-shaoge — APPROVED** [review]

**Pull request merged**